pkg/datapath/linux/node.go

```diff
@@ -1081,26 +1081,29 @@ func (n *linuxNodeHandler) enableIPsecIPv4(newNode *nodeTypes.Node, nodeID uint1
 	var spi uint8
 	var errs error
 
-	if !n.nodeConfig.EnableIPv4 || newNode.IPv4AllocCIDR == nil {
+	if !n.nodeConfig.EnableIPv4 || (newNode.IPv4AllocCIDR == nil && !n.subnetEncryption()) {
 		return nil
 	}
 
-	new4Net := newNode.IPv4AllocCIDR.IPNet
 	wildcardIP := net.ParseIP(wildcardIPv4)
 	wildcardCIDR := &net.IPNet{IP: wildcardIP, Mask: net.IPv4Mask(0, 0, 0, 0)}
 
 	err := ipsec.IPsecDefaultDropPolicy(false)
 	errs = errors.Join(errs, upsertIPsecLog(err, "default-drop IPv4", wildcardCIDR, wildcardCIDR, spi))
 
 	if newNode.IsLocal() {
-		errs = errors.Join(errs, n.replaceNodeIPSecInRoute(new4Net))
```

Comment:
IIUC, on EKS & AKS, we would install a route using the internal auto-generated IPv4/IPv6AllocRange. Do we have an idea of the potential consequences of that? The second commit suggests that the auto-generated CIDR could clash with real CIDR, although unlikely.

Reply:
So yes, before this PR installed a route for the auto-generated pod CIDR. It used an auto-generated pod CIDR, because that's Cilium's ancient fall-back if no pod CIDR is configured. That auto-generated pod CIDR is not used anywhere (see also https://github.com/cilium/cilium/issues/27018) and we're planning to eventually remove it, since it causes confusion for both engineers and users.

Particularly in EKS and AKS, IPs are not allocated from any pod CIDR (be it auto-generated or manually configured), and so using a node's pod CIDR in AKS or EKS is making assumptions about the IP range that does not hold true. By installing a route for each node's (auto-generated) pod CIDR (as we did before the PR), we routed all traffic belonging to those auto-generated CIDRs to the encrypt device.  

I'm not sure, what the consequence of this were. We have not received bug reports about this, I think mostly because the auto-generated CIDR is by design unlikely to clash with a real CIDR. Since the encrypt device is probably also the default egress device, the route might also not have affected any issues for CIDRs that were routable on the default egress device anyway.

In any case, my primary motivation for this PR is the clean up unconditional uses of "pod CIDR" in the agent code, because such code will not be compatible with pod-CIDR-less IPAM modes such as EKS/AKS or Multi-Pool. IPSec already has "subnet encryption" as a pod CIDR replacement, but for some reason it still unconditionally used the pod CIDR, which this PR attempts to fix.

Reply:
Awesome, thanks for that test :rocket: 

> I have no idea how unlikely it is for the CIDRs to clash, but maybe we need to add a release note (something like below or can we make the "potential" more precise?) and backport to just v1.14?

In the end, it depends on user configuration. [The auto-generated CIDR](https://github.com/cilium/cilium/blob/a8d80e25ca2d26affe4f7f7a8350bb537ddbe388/pkg/node/address.go#L123) is of the form "10.x.0.0/16" where x is derived from the node's IP. So I think clashes can happen - but I'm not aware of any bug reports, probably because most users are using IPs in the range of "10.0.0.0/16".

I think we can absolutely backport this to v1.14. I removed the backport label because there have not been bug reports and this has likely been present for years, but given the risk should be low, it does fit our backporting policy.